src/Microsoft.ML.LightGBM/LightGbmTrainerBase.cs

```diff
@@ -0,0 +1,883 @@
+        }
+
+        protected virtual void CheckDataValid(IChannel ch, RoleMappedData data)
+        {
+            data.CheckFeatureFloatVector();
+            ch.CheckParam(data.Schema.Label != null, nameof(data), "Need a label column");
+        }
+
+        protected virtual void GetDefaultParameters(IChannel ch, int numRow, bool hasCategarical, int totalCats, bool hiddenMsg=false)
+        {
+            double learningRate = Args.LearningRate ?? DefaultLearningRate(numRow, hasCategarical, totalCats);
+            int numLeaves = Args.NumLeaves ?? DefaultNumLeaves(numRow, hasCategarical, totalCats);
+            int minDataPerLeaf = Args.MinDataPerLeaf ?? DefaultMinDataPerLeaf(numRow, numLeaves, 1);
+            Options["learning_rate"] = learningRate.ToString();
+            Options["num_leaves"] = numLeaves.ToString();
+            Options["min_data_per_leaf"] = minDataPerLeaf.ToString();
+            if (!hiddenMsg)
+            {
+                if (!Args.LearningRate.HasValue)
+                    ch.Info("Auto-tuning parameters: " + nameof(Args.LearningRate) + " = " + learningRate);
+                if (!Args.NumLeaves.HasValue)
+                    ch.Info("Auto-tuning parameters: " + nameof(Args.NumLeaves) + " = " + numLeaves);
+                if (!Args.MinDataPerLeaf.HasValue)
+                    ch.Info("Auto-tuning parameters: " + nameof(Args.MinDataPerLeaf) + " = " + minDataPerLeaf);
+            }
+        }
+
+        private FloatLabelCursor.Factory CreateCursorFactory(RoleMappedData data)
+        {
+            var loadFlags = CursOpt.AllLabels | CursOpt.AllWeights | CursOpt.Features;
+            if (_predictionKind == PredictionKind.Ranking)
+                loadFlags |= CursOpt.Group;
+
+            var factory = new FloatLabelCursor.Factory(data, loadFlags);
+            return factory;
+        }
+
+        private static List<int> GetCategoricalBoundires(int[] categoricalFeatures, int rawNumCol)
+        {
+            List<int> catBoundaries = new List<int> { 0 };
+            int curFidx = 0;
+            int j = 0;
+            while (curFidx < rawNumCol)
+            {
+                if (j < categoricalFeatures.Length && curFidx == categoricalFeatures[j])
+                {
+                    if (curFidx > catBoundaries.Last())
+                        catBoundaries.Add(curFidx);
+                    if (categoricalFeatures[j + 1] - categoricalFeatures[j] >= 0)
+                    {
+                        curFidx = categoricalFeatures[j + 1] + 1;
+                        catBoundaries.Add(curFidx);
+                    }
+                    else
+                    {
+                        for (int i = curFidx + 1; i <= categoricalFeatures[j + 1] + 1; ++i)
+                            catBoundaries.Add(i);
+                        curFidx = categoricalFeatures[j + 1] + 1;
+                    }
+                    j += 2;
```

Comment:
>j [](start = 20, length = 1)

out of curiosity, why j? why not i?

Reply:
j is just fine.

---
In reply to: [197551532](https://github.com/dotnet/machinelearning/pull/392#discussion_r197551532) [](ancestors = 197551532)